dcp.1
.\" @(#)man page for netl network logger/sniffer
.\" copyright 1997 graham the ollis
.ad b
.TH dcp 1 "3 june 1997" "CORE software" "CORE software"
.AT 3
.de sh
.br
.ne 5
.PP
\fB\\$1\fR
.PP
..
.SH NAME
dcp \- discrete comunication protocol
.SH SYNOPSIS
.B dcp
mesage [port]
.SH DESCRIPTION
.I dcp
is a way of passing messages to 
.BR netl (8)
and related programs.  in the case of 
.BR netl (8)
there is currently only one message you can send it.  "netl:readconfig" will
order 
.BR netl (8)
(if it's listening, see the man page for details on the 
.I listen
directive) to terminate it's current configeration and read the 
netl.conf file again.  this is handy if you make modifications to netl.conf
while 
.BR netl (8)
is running.
.SH OPTIONS
none
.SH SEE ALSO
.BR neta (1),
.BR xd (1),
.BR netl (8),
.BR hwpassive (8)
.SH BUGS
almost certainly.  please send bug reports/fixes to me 
Graham THE Ollis <ollisg@ns.arizona.edu>
.SH COPYING
Copyright 1996, 1997 Graham THE Ollis
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.SH AUTHOR
Graham THE Ollis <ollisg@ns.arizona.edu>


HISTORY
version .95.2
-------------

 . mostly internal stuff.  added parse.c and grab.c modules, which contain
   code that used to be in the netl.c main module.  this is so that all the 
   network specific code is in grab.c, the true nightmare for porting is
   in this file... so if i can port grab.c to win32... then i'm done.

 . the \ character now works in the config file, if you want to have
   multiple line rules.  very unix-esq.

version .95
-----------

 . created man page documentation.

 . added `detect' config line, see the netl(8) manpage for details.
   this will reduce the amount of configeration you will have to do
   in the netl.conf file.

 . modified dist version of netl.conf so that it's more portable.
   in particular, i put in literal port numbers rather than using
   port names.

 . dcp.pl uses perl native Socket module instead of using net cat.
   `dcp.pl -o' uses the old method.

version .94
-----------

 . changed all the struct declerations to typedef declerations.  get off my
   back ok?  i've been programming perl for so long, forgot some of that
   strange C language. 

 . no longer depending on linux .h files for IP (and sub) headers.  this
   is for portability reasons.  if you are doing a port, make sure you get
   the endian type right!  there are now three global headers:
	global.h - general global stuff (endian integer conversion etc.)
	ether.h - ethernet specific data types, #defines
	ip.h - ip protocol specific data types, #defines
   logically, if your adding an interface (ATM, token ring, ppp, etc.)
   you'd want to start in ether.h.  if your adding an OS, look at global.h
   first.  if you want to add a protocol (IPX, etc.) start with ip.h and
   modify/add to these files.  if your trying to do all of this at once,
   god help you. 

 . fixed the clearipcache() function.  thus now, when you send a 
   netl:readconfig, netl will clear the ip cache before reading the file,
   avoiding all sorts of potential duplicates.  though you probably have
   an enourmous amount of memory like me and don't care... 

 . added some global options with ports to win32 in mind.  NO_SYSLOGD, if 
   defined in global.h will not even compile with the syslogd library stuff.  
   this is a good idea if your running from a non unix system.  note that 
   syslogd is really nice in that it partially compresses lines by indicating 
   repeated lines.  this won't save you if you get strobed, but will keep your 
   log file neat if some one ping -f's you from only one site.

 . added NO_TEEOUT option which is defined for unix.  the -o option should 
   only be necessary in win32 only.  even then, you should only compile with
   this option if your thinking of using it.  -o copies output to the specified
   file.

0.93 (and 0.92)
---------------

 a. -z option allows you to run netl in the foreground.  output is sent to
    stdout instead of syslogd.

 b. you can include configeration lines in the command line.  when you do
    this, the config file /etc/netl.conf will NOT be read.  note that (a)
    and (b) are meant to be used in concert with each other for testing
    and quick monitoring. 

 c. dcp is partially implemented.  netl will listen to port 47 udp
    (silently) if you `listen 47' in your config file.  when you send the
    message `netl:readconfig' netl is suppose to re read your config file. 
    dcp.pl is a perl script (you'll need net cat for this to work) which
    sends this message.  what is left to do for dcp?
	1. generic server side dcp interface libdcp.a file or somesuch.
	2. more solid C version of dcp client.

 d. netl is now officially under the GPL, see COPYING for more information.

 e. icmp codes are now included in some .h file somewhere rather than
    depending on your linux being overly up to date.

 f. neta no longer swaps source/destination ip/ethernet addresses
    inapropriately. 

 g. started work on hwpassive, which is no where near complete.

 h. tdr.pl is no included with netl distribution for convenience.  get
    your own version at: 

http://www.u.arizona.edu/~ollisg/core.html
	[ CORE ]

netl 0.91
---------

 minor updates

netl 0.90
---------

 initial release


dcp.c
/*==============================================================================
| dcp.c - discrete comunication protocol
|   optimized (and debugged) by Graham THE Ollis <ollisg@ns.arizona.edu>
|
|   Copyright (C) 1997 Graham THE Ollis <ollisg@ns.arizona.edu>
|
|   This program is free software; you can redistribute it and/or modify
|   it under the terms of the GNU General Public License as published by
|   the Free Software Foundation; either version 2 of the License, or
|   (at your option) any later version.
|
|   This program is distributed in the hope that it will be useful,
|   but WITHOUT ANY WARRANTY; without even the implied warranty of
|   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
|   GNU General Public License for more details.
|
|   You should have received a copy of the GNU General Public License
|   along with this program; if not, write to the Free Software
|   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
|
|  Date       Name	Revision
|  ---------  --------  --------
|  09 Mar 97  G. Ollis	created module
|=============================================================================*/

#include <string.h>
#include <stdio.h>
#include <netinet/in.h>

#include "global.h"
#include "ether.h"
#include "ip.h"

#include "io.h"
#include "config.h"
#include "dcp.h"
#include "options.h"
#include "resolve.h"

/*==============================================================================
| globals
|=============================================================================*/

u32 lasthearid = 0;

/*==============================================================================
| soundwave: jam that transmission!
| process a comunication request
|=============================================================================*/

void
hear(u8 *dg, udphdr *h, int len)
{
  char		message[MAX_COM_LEN];
  int		size, offset;
  u32		id;
  u16		nsize;

  /*============================================================================
  | convert the udp packet in to a c string.
  | that's c string as in STANDARD ANSI C character array type STRING thing.
  | not "i'm a sorry ass microsoft lacky who's going to spend the rest of his
  | life playing silly games while graham takes over the world" c strings.
  |===========================================================================*/

  offset = sizeof(machdr) + sizeof(iphdr) + sizeof(udphdr);

  id = ntohl(*((u32 *) &dg[offset]));		offset += 4;
  if(id == lasthearid)
    return;
  lasthearid = id;
  nsize = ntohs(*((u16 *) &dg[offset]));	offset += 2;

  size = len - offset;
  if(nsize < size)
    size = nsize;

  if(size > MAX_COM_LEN - 1)
    size = MAX_COM_LEN - 1;
  memcpy(message, &dg[offset], size);
  message[size] = '\0';

  /*============================================================================
  | all comunication requests are logged, even if we ignore them
  |===========================================================================*/

  log("dcp[%d]: \"%s\"", id, message);

  if(!strncmp("netl:", message, 5)) {
    if(!strncmp("readconfig", &message[5], 10)) {

      clearipcache();
      log("old ip cache cleared");
     
      clearconfig();
      log("old configuration cleared");

      preconfig();
#ifdef NO_SYSLOGD
      readconfig(configfile);
#else
      readconfig(configfile, noBackground);
#endif
      postconfig();
      log("re-read configfile %s", configfile);

    } else 
      err("warning: unknown netl comunication request %s", &message[5]);
  }
}

